### PR TITLE
JDK-8280498: jdk/java/net/Inet4Address/PingThis.java fails on AIX

### PR DESCRIPTION
**Pull request by aixtools**
with IP "0.0.0.0"

- it either does nothing and ping fails, or, in some virtual environments
is treated as the default route address.
- IPv6 support for ::1 is available since 1977; however, ::0 is not accepted
as a vaild psuedo IPv6 address. '::1' must be used instead.
```
ping: bind: The socket name is not available on this system.
ping: bind: The socket name is not available on this system.
PING ::1: (::1): 56 data bytes
64 bytes from ::1: icmp_seq=0 ttl=255 time=0.037 ms
64 bytes from ::1: icmp_seq=1 ttl=255 time=0.045 ms

--- ::1 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0/0/0 ms
PING ::1: (::1): 56 data bytes
64 bytes from ::1: icmp_seq=0 ttl=255 time=0.052 ms
64 bytes from ::1: icmp_seq=1 ttl=255 time=0.047 ms

--- ::1 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
```

A long commit message. 

This came to light because some systems failed with IPv4 (those that passed
replaced 0.0.0.0 with the default router. but most just fail - not substituting
0.0.0.0 with 127.0.0.1. However, InetAddress.getByName("") returns 127.0.0.1
which compares well with other platform's behavior.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8280498](https://bugs.openjdk.java.net/browse/JDK-8280498): jdk/java/net/Inet4Address/PingThis.java fails on AIX

### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)
 * [Michael McMahon](https://openjdk.java.net/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)
 * [Aleksei Efimov](https://openjdk.java.net/census#aefimov) (@AlekseiEfimov - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7013/head:pull/7013` \
`$ git checkout pull/7013`

Update a local copy of the PR: \
`$ git checkout pull/7013` \
`$ git pull https://git.openjdk.java.net/jdk pull/7013/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7013`

View PR using the GUI difftool: \
`$ git pr show -t 7013`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7013.diff">https://git.openjdk.java.net/jdk/pull/7013.diff</a>

</details>

### CODE REVIEW

**Comment by tstuefe** [comment]
Hi @aixtools,

Welcome! Good job with aixtools.net, we used it quite a bit over the years :)

About your patch, probably it will be ignored in this form. Some notes:
- you need an official JBS issue for the toolchain to notify relevant mailing lists. I can open one for you if you can give me a short concise bug text explaining the issue.
- we have a ppc/aix mailing list. It's a bit deserted, but still a good place to ask around or notify ppl of your intent to patch. https://mail.openjdk.java.net/mailman/listinfo/ppc-aix-port-dev
- and of course, we need you to sign the OCA.

Officially, I think IBM is now maintainer of the AIX port (they took over from us, SAP). Are you affiliated with IBM?

Cheers, Thomas

**Comment by aixtools** [comment]
I was affiliated with IBM. I am more affiliated with Adoptium (was 
OpenJDK) these days.

However, as most of the work that IBM will be doing (that they took over 
from you) is hosted at the same location - I do assist IBM with the 
basic setup and config.

The testing (and related PR's) comes from the testing done by Adoptium.

On 18/01/2022 07:29, Thomas Stuefe wrote:
>
> Hi @aixtools <https://github.com/aixtools>,
>
> Welcome! Good job with aixtools.net, we used it quite a bit over the 
> years :)
>
a bit sad to see 'used' (past tense) - but very glad I could help. I 
hate, well, try VERY VERY hard to avoid, unneeded dependencies.

I am told my work on github/cpython (helped) convinced IBM to use XLC as 
compiler for Python on AIX 7.3 - so that is a step forward (imho).

> About your patch, probably it will be ignored in this form. Some notes:
>
>   * you need an official JBS issue for the toolchain to notify
>     relevant mailing lists. I can open one for you if you can give me
>     a short concise bug text explaining the issue.
>
Would appreciate that: I'll get back on the consise text. Have to find 
my C program I used for testing the library call. From memory, an 
argument (""), i.e., null-length string behaves as expected for 
"0.0.0.0", whereas (NULL) does not (fails iirc), and "0.0.0.0" is taken 
as a valid address - that it fails to resolve.

And, also, "::0" is simply refused (also by ping on CLI), so to get that 
to work as expected ping to "::1" - which has been available since 1997 
(iirc, when AIX 4.3.0 was released with dual stack support (i.e., "::1" 
does not need `autoconf6` to be called to be available, it is always there).

So, you conld consider it a bug that the AIX library does not properly 
handle NULL (as documented, iirc), and certainly an inconvience that 
"0.0.0.0" is not handled the same way by the library function - compared 
to the "ping" command (where some environments transform "0.0.0.0" to 
the IP of the default router. Not been able to figure out how that 
happens (or where) - but I think some routers might be responding to 
"0.0.0.0" as "here" as was stated in the ancient docs re: meaning of 
"0.0.0.0".

If above is sufficient for the JBS issue - fantistic - otherwise I'll 
try and get something less chatty from code.

>   * we have a ppc/aix mailing list. It's a bit deserted, but still a
>     good place to ask around or notify ppl of your intent to patch.
>     https://mail.openjdk.java.net/mailman/listinfo/ppc-aix-port-dev
>
registering...
>
>   * and of course, we need you to sign the OCA.
>
where is that?

**Comment by tstuefe** [comment]
>If above is sufficient for the JBS issue - fantistic - otherwise I'll
try and get something less chatty from code.

That would be nice.

>> * and of course, we need you to sign the OCA.
> where is that?

https://openjdk.java.net/contribute/ outlines the process. Point 0 is needed, signing the Oracle contributor agreement. Note that I am not affiliated with Oracle. It's what they require.

**Comment by tstuefe** [comment]
I opened an issue for you to track this.

[JDK-8280498: [aix]: jdk/java/net/Inet4Address/PingThis.java fails
](https://bugs.openjdk.java.net/browse/JDK-8280498)

Please change the title of this PR to "JDK-8280498: [aix]: jdk/java/net/Inet4Address/PingThis.java fails" to allow our bots to make the connection.

**Comment by RealCLanger** [comment]
Hi, @aixtools,
you will also first have to take care for the oca. Before that state isn't clear, none of the discussions on this PR will be forwarded to the mailing lists.

**Comment by aixtools** [comment]
Does the label being removed mean the site knows I now have an _oca_?

**Comment by tstuefe** [comment]
> Does the label being removed mean the site knows I now have an _oca_?

Yes. The oca label was removed by the bot. It also did sent the mail to the associated ML (https://mail.openjdk.java.net/pipermail/net-dev/2022-January/017386.html).

You still have no write access to the JBS issue, because you need "authorship" for that. But the bar is pretty low (I believe three sponsored changes or so).

Sorry for the complexity. It's all a tradeoff - attracting developers vs keeping noise down.

As soon as someone reviews, I can sponsor the change.

Cheers, Thomas

File: test/jdk/java/net/Inet4Address/PingThis.java [hunk 1]
```diff
@@ -48,15 +48,26 @@ public static void main(String args[]) throws Exception {
...
         List<String> addrs = new ArrayList<String>();
 
         if (IPSupport.hasIPv4()) {
-            addrs.add("0.0.0.0");
+            if (isAIX) {
+                addrs.add("");
+            }
+            else {
+                addrs.add("0.0.0.0");
+            }
```
Comment (AlanBateman) [inline test/jdk/java/net/Inet4Address/PingThis.java:62]: This conflicts with the purpose of the test. Maybe this test needs to be skipped on AIX?

Comment (Michael-Mc-Mahon) [reply test/jdk/java/net/Inet4Address/PingThis.java:62]: Or is it possible to change the implementation on AIX so the test passes without change?

Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:62]: If it defeats the purpose, then it needs to be skipped.

* When I was trying to understand the test, it seemed to be that it _assumed_ that "0.0.0.0" was 'converted' to 127.0.0.1.
* If there is an international standard (ISO, POSIX) that states 0.0.0.0 needs to be treated as 127.0.0.1 - AIX clearly has a bug. Same for `::)` mapping to `::1` - AIX doesn't do that IPv4 (now), and, afaik, never has for IPv6.
* While testing I also saw that AIX reacts differently (error message iirc) when passed NULL, whereas it behaves "as expected" when given a null-string `""`
* FYI: Adoptium has put this test on the excluded list, but if a fix using `""` rather than "0.0.0.0" and/or "::1" rather than "::0", the test should perhaps state the exclusion as a comment in the code, and return immediately, as it does for Windows.

**Comment by aixtools** [comment]
> > Does the label being removed mean the site knows I now have an _oca_?
> 
> Yes. The oca label was removed by the bot. It also did sent the mail to the associated ML (https://mail.openjdk.java.net/pipermail/net-dev/2022-January/017386.html).
> 
> You still have no write access to the JBS issue, because you need "authorship" for that. But the bar is pretty low (I believe three sponsored changes or so).
> 
> Sorry for the complexity. It's all a tradeoff - attracting developers vs keeping noise down.
Basically, it nearly kept from submitting anything at all.
> 
> As soon as someone reviews, I can sponsor the change.
Thx.
> 
> Cheers, Thomas

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 1 above)
Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:62]: > Or is it possible to change the implementation on AIX so the test passes without change?

* Digging into the java guts to map "0.0.0.0" to "127.0.0.1" seems too far to me - as I believe interfaces are not suppossed to be "translating" as well.
* If the behavior "0.0.0.0" becomes "127.0.0.1" is an official standard - a bug needs to be filed with IBM (which I cannot do) - but we need to also be aware that any fix to AIX is not likely to ever be applied on AIX 7.1 TL4 (maybe TL5) which are systems used to build distributions.
* Again - being a noob (or fresh face; fresh meat) the test title is merely "PingThis" - where is "This" defined as "0.0.0.0" (if it had been named _PingHERE_ I could understand as the alias, iirc, for 0.0.0.0 is "Here".

Comment (Michael-Mc-Mahon) [reply test/jdk/java/net/Inet4Address/PingThis.java:62]: Might be better to skip the test then by adding to the header:

 * 
 * @requires os.family != "aix"
 * 
rather than the check in the test source

Comment (dfuch) [reply test/jdk/java/net/Inet4Address/PingThis.java:62]: From the description and evaluation of  [JDK-7163874](https://bugs.openjdk.java.net/browse/JDK-7163874) it seems that skipping this test on AIX would be reasonable.

Some operating systems seem to accept 0.0.0.0 as input - and reply with 127.0.0.1. The changes made for  JDK-7163874 simply makes sure that such a reply is not rejected as invalid on the ground that the received address is not the same as the input address. If 0.0.0.0 is not a valid input for the underlying operating system, then the test has no object and should probably be skipped.

**Comment by aixtools** [comment]
p.s. Just want to be clear: On AIX
- ping 0.0.0.0 is permissible, but responds differently
- ping ::0 is not supported
```
aixtools@x064:[/home/aixtools]ping 0.0.0.0
PING 0.0.0.0: (0.0.0.0): 56 data bytes

--- 0.0.0.0 ping statistics ---
3 packets transmitted, 0 packets received, 100% packet loss
aixtools@x064:[/home/aixtools]ping ::0
ping: bind: The socket name is not available on this system.
aixtools@x064:[/home/aixtools]ping ::1
PING ::1: (::1): 56 data bytes
64 bytes from ::1: icmp_seq=0 ttl=255 time=0.053 ms
64 bytes from ::1: icmp_seq=1 ttl=255 time=0.095 ms

--- ::1 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
```
and
```
aixtools@gcc119:[/home/aixtools]ping 0.0.0.0
PING 0.0.0.0: (0.0.0.0): 56 data bytes
64 bytes from 140.211.9.1: icmp_seq=0 ttl=255 time=0 ms
64 bytes from 140.211.9.1: icmp_seq=1 ttl=255 time=0 ms
64 bytes from 140.211.9.1: icmp_seq=2 ttl=255 time=0 ms

--- 0.0.0.0 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0/0/0 ms
aixtools@gcc119:[/home/aixtools]ping ::0
ping: bind: The socket name is not available on this system.
aixtools@gcc119:[/home/aixtools]ping ::1
PING ::1: (::1): 56 data bytes
64 bytes from ::1: icmp_seq=0 ttl=255 time=0.045 ms
64 bytes from ::1: icmp_seq=1 ttl=255 time=0.080 ms

--- ::1 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0/0/0 ms
```

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 1 above)
Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:62]: > From the description and evaluation of [JDK-7163874](https://bugs.openjdk.java.net/browse/JDK-7163874) it seems that skipping this test on AIX would be reasonable.
> 
> Some operating systems seem to accept 0.0.0.0 as input - and reply with 127.0.0.1. The changes made for JDK-7163874 simply makes sure that such a reply is not rejected as invalid on the ground that the received address is not the same as the input address. If 0.0.0.0 is not a valid input for the underlying operating system, then the test has no object and should probably be skipped.

Being a bit more specific - especially re: valid address:
- On AIX ping ::0 is not valid
```
aixtools@x064:[aixtools]ping ::0
ping: bind: The socket name is not available on this system.
- On AIX ping 0.0.0.0 is not invalid, but how it responds
```
ETH: ====( 98 bytes transmitted on interface en0 )==== 16:13:13.970131171
ETH:    [ 00:21:5e:a3:c7:44 -> 44:4e:6d:fa:4b:fe ]  type 800  (IP)
IP:     < SRC =  192.168.129.64 >  (x064)
IP:     < DST =         0.0.0.0 >
IP:     ip_v=4, ip_hl=20, ip_tos=0, ip_len=84, ip_id=63721, ip_off=0
IP:     ip_ttl=255, ip_sum=80d6, ip_p = 1 (ICMP)
ICMP:   icmp_type=8 (ECHO_REQUEST)  icmp_id=216  icmp_seq=516
```
- In special cases - when there are devices external to the (virtual) machine, that respond to 0.0.0.0 you get the address of the replying device:
```
root@p8-java1-adopt07:[/root]ping 0.0.0.0
PING 0.0.0.0 (0.0.0.0): 56 data bytes
64 bytes from 140.211.9.1: icmp_seq=0 ttl=255 time=11 ms
64 bytes from 140.211.9.1: icmp_seq=1 ttl=255 time=12 ms
64 bytes from 140.211.9.1: icmp_seq=2 ttl=255 time=0 ms
```
- At trace level:
```
ETH: ====( 98 bytes transmitted on interface en1 )==== 16:19:39.475020041
ETH:    [ 12:31:9e:0f:73:03 -> 58:bc:27:ae:b0:50 ]  type 800  (IP)
IP:     < SRC =   140.211.9.168 >  (p8-java1-adopt07)
IP:     < DST =         0.0.0.0 >
IP:     ip_v=4, ip_hl=20, ip_tos=0, ip_len=84, ip_id=23686, ip_off=0
IP:     ip_ttl=255, ip_sum=c8a7, ip_p = 1 (ICMP)
ICMP:   icmp_type=8 (ECHO_REQUEST)  icmp_id=238  icmp_seq=18

ETH: ====( 98 bytes received on interface en1 )==== 16:19:39.475356474
ETH:    [ 58:bc:27:ae:b0:50 -> 12:31:9e:0f:73:03 ]  type 800  (IP)
IP:     < SRC =     140.211.9.1 >  (corv-car1-gw.nero.net)
IP:     < DST =   140.211.9.168 >  (p8-java1-adopt07)
IP:     ip_v=4, ip_hl=20, ip_tos=0, ip_len=84, ip_id=23686, ip_off=0
IP:     ip_ttl=255, ip_sum=32d3, ip_p = 1 (ICMP)
ICMP:   icmp_type=0 (ECHO_REPLY)  icmp_id=238  icmp_seq=18
```

In conclusion: the `solution` is most likely to skip the test, just as it is skipped for Windows.
a) address `::0` is not valid
b) a response to IPv4: 0.0.0.0 depends on an external device responding to `0.0.0.0`

I'll make the basic changes - but I am sure to need help with text (the right bug number, or whatever) that needs to go into the code.

**Comment by aixtools** [comment]
Force pushed - too many typos - so 'erased' the mess-ups

File: test/jdk/java/net/Inet4Address/PingThis.java [hunk 2]
```diff
@@ -32,6 +32,9 @@
  *          for InetAdress 0.0.0.0 and ::0
  * @run main PingThis
  * @run main/othervm -Djava.net.preferIPv4Stack=true PingThis
+ * @bug 8280498
+ * @summary InetAddress ::0 is not valid, InetAddress 0.0.0.0 only responds
+ *          when an external device reacts to 0.0.0.0
```
Comment (AlanBateman) [inline test/jdk/java/net/Inet4Address/PingThis.java:37]: I think this means the test has two @summary tags, did you mean to add a second?

**Comment by dfuch** [comment]
@aixtools Thanks for the detailed analysis. It confirms the best course of action is to skip this test on AIX. Since changes will be to the test file only there's no need to list  8280498 in the test's `@bug` list - adding the `noreg-self` keyword to JDK-8280498 would be enough (I can do that).

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 2 above)
Comment (dfuch) [inline test/jdk/java/net/Inet4Address/PingThis.java:37]: Maybe that information could be added as a comment just before `return;` on line 52 instead:
`// On AIX, ...`

**Comment by aixtools** [comment]
Deleted PR because the test is being skipped (at Adoptium).

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 2 above)
Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:37]: I removed the @summary bit, as well as the reference to the bug, and moved the comments. However, I did not use // to start the comments - as I do not like 'mixed' comment formatting.

**Review by dfuch — APPROVED** [review]

File: test/jdk/java/net/Inet4Address/PingThis.java [hunk 3]
```diff
@@ -45,7 +45,12 @@
 
 public class PingThis {
     public static void main(String args[]) throws Exception {
-        if (System.getProperty("os.name").startsWith("Windows")) {
+        osname = System.getProperty("os.name")
+        /*
```
Comment (Michael-Mc-Mahon) [inline test/jdk/java/net/Inet4Address/PingThis.java:49]: Doesn't look like the line above will compile.

**Review by Michael-Mc-Mahon — CHANGES_REQUESTED** [review]

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 3 above)
Comment (dfuch) [reply test/jdk/java/net/Inet4Address/PingThis.java:49]: Good catch!

Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:49]: too much Python (no semi-colons) - great catch.

**Review by AlekseiEfimov — APPROVED** [review]
The changes looks ok to me given that the test can be compiled and executed with no failures. 
The following uestion being asked before ,but is there a reason why we check platform in the test code, instead of using [requires jtreg tag](https://openjdk.java.net/jtreg/tag-spec.html#requires_names) ?  
The following requires tag could be used as a replacement for the proposed change:
```
* @requires (os.family != "windows" & os.family != "aix")
```

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 3 above)
Comment (jaikiran) [reply test/jdk/java/net/Inet4Address/PingThis.java:49]: @aixtools, if you enable GitHub Actions on your forked repo, it should then automatically trigger a GitHub Actions job for your changes in your PR and those results/run will be visible in this PR. The openjdk/jdk repo has been configured to run `tier1` build/test on PR submission, so errors like this are easily caught. Of course, if you build this locally, that's much quicker to catch too.

Comment (dfuch) [reply test/jdk/java/net/Inet4Address/PingThis.java:49]: This is a test in tier2 though - so I don't expect it will be caught by github actions

**Comment by aixtools** [comment]
@AlekseiEfimov - that is, imho, a more elegant way to skip tests. I'll look for an example elsewhere and make the change.

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 3 above)
Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java]: re: actions - the default seems to be enable all, so unclear on what I would need to do extra to get these working.
Guess I need to study how to build locally, as well as how to use the 'custom' variation of a Grinder (jenkins) test.

(I'm just an `admin`, lets say the 'ops` side of `devops` - so I have many things to learn re: the `dev` side of things).

Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java]: Saying resolved - after using @requires rather than check in code - removing the windows code check as well

**Comment by aixtools** [comment]
@AlekseiEfimov , @dfuch - please review again.

File: test/jdk/java/net/Inet4Address/PingThis.java [hunk 4]
```diff
@@ -27,9 +27,14 @@
 
 /* @test
  * @bug 7163874 8133015
+ * @requires os.family != "windows" && os.family != "aix"
  * @library /test/lib
  * @summary InetAddress.isReachable is returning false
```
Comment (Michael-Mc-Mahon) [inline test/jdk/java/net/Inet4Address/PingThis.java:32]: '&&' double ampersand syntax does not work here. Should be a single ampersand as suggested by Aleksei.

**Review by Michael-Mc-Mahon — CHANGES_REQUESTED** [review]

File: test/jdk/java/net/Inet4Address/PingThis.java [hunk 5]
```diff
@@ -27,9 +27,14 @@
...
  *          for InetAdress 0.0.0.0 and ::0
```
Comment (AlekseiEfimov) [inline test/jdk/java/net/Inet4Address/PingThis.java:33]: Typo: `InetAdress` -> `InetAddress`

File: test/jdk/java/net/Inet4Address/PingThis.java [hunk 6]
```diff
@@ -27,9 +27,14 @@
...
+ *          On AIX InetAddress ::0 is not valid, InetAddress 0.0.0.0 is valid,
+ *          but does not respond as 127.0.0.1 (i.e., only responds
+ *          when an external device reacts to 0.0.0.0)
+ */
```
Comment (AlekseiEfimov) [inline test/jdk/java/net/Inet4Address/PingThis.java:37]: This line should be removed to fix formatting of the comment section, ie unbalanced `*/`

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 4 above)
Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:32]: Once a C programmer, hard to resist. Thx. :)

File: test/jdk/java/net/Inet4Address/PingThis.java (hunk 6 above)
Comment (aixtools) [reply test/jdk/java/net/Inet4Address/PingThis.java:37]: Thx!

**Review by Michael-Mc-Mahon — APPROVED** [review]
LGTM

**Review by AlekseiEfimov — APPROVED** [review]
Latest changes look good to me. Will check that the test runs/not runs with our CI and will sponsor it once I have the results.

**Comment by aixtools** [comment]
- Thx all for your assistance and patience.
- (Hoping you will consider this for backport as well).

**Comment by AlekseiEfimov** [comment]
Hi @aixtools,
The test runs fine on non-windows/non-aix platforms. The fix is ready for the integration: I will sponsor it after you issue `/integrate` [command](https://github.com/openjdk/jdk/pull/7013#issuecomment-1111165207).

**Review by dfuch — APPROVED** [review]
The last version look good to me too. I agree that using `@requires` is a cleaner way to skip or select the test. Thanks for making this change!

**Comment by AlekseiEfimov** [comment]
Hi @aixtools,
